sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,814 @@
+`usm::alloc::host` or `usm::alloc::unknown`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(queue q, usm::alloc kind, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool which is associated with the context and
+device from the queue `q``, with the allocation kind `kind` and applying any
+properties in `props`. If `kind` is `usm::alloc::host` this is equivalent to
+calling `memory_pool(queue.get_context(), props)` otherwise this is equivalent
+to calling `memory_pool(queue.get_context(), queue.get_device(), props)`.
+
+_Throws_: An exception with the `errc::invalid` error code if `kind` is
+`usm::alloc::unknown`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, void *ptr, size_t size, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool which is associated with context `ctx`, 
+uses an existing USM memory allocation `ptr` of size (in bytes) `size`,
+and applying any properties in `props`. The memory pool will use the existing
+USM memory allocation instead of allocating from the SYCL runtime. The
+associated device and allocation kind of the memory pool is inferred from the
+`ptr`. If `ptr` is not a valid USM pointer of kind `usm::alloc::host`,
+`usm::alloc::device` or `usm::alloc::shared` or the value `size` does not match
+the size of memory allocated for `ptr` the result is undefined behavior. If
+`ptr` is freed or the memory it points to is written to whilst the memory pool
+is still alive the result is undefined behavior. The maximum size and initial
+threshold of the memory pool are set to `size`.
+
+_Throws_: An exception with the `errc::invalid` error code if `props` contains
+the `maximum_size` or `initial_threshold` properties.
+
+
+[source, c++]
+----
+~memory_pool();
+----
+
+_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
+pool to be destroyed after all remaining allocations have been freed, and
+returns immediately without waiting.
+
+[source, c++]
+----
+context get_context() const;
+----
+
+_Returns_: The SYCL context associated with the memory pool.
+
+[source, c++]
+----
+device get_device() const;
+----
+
+_Returns_: The SYCL device associated with the memory pool. If the memory pool
+allocates memory of kind `usm::alloc::device` or `usm::alloc::host::shared` the
+device returned is the same device which was passed when constructing the memory
+pool. If the memory pool allocates memory of kind `usm::alloc::host` the device
+returned is the first of the devices associated with the context, which is
+associated with the memory pool.
+
+[source, c++]
+----
+usm::alloc get_alloc_kind() const;
+----
+
+_Returns_: The memory allocation kind of the memory pool.
+
+[source, c++]
+----
+size_t get_allocation_chunk_size() const;
+----
+
+_Returns_: The allocation chunk size of the device associated with the memory
+pool.
```

Comment:
Is it important to provide an API to query the chunk size?  It seems like this is an implementation detail.  The only affect on the API is with `set_new_threshold` and `maximum_size`.  It seems like `set_new_threshold` does not need to say precisely how the threshold value is rounded because the threshold is just a hint anyways.  For `maximum_size`, it seems like the wording could say something like:

> Because the memory pool may allocate memory in discrete chunks, the value may be rounded up to account for the size of these chunks.

It does not seem necessary for the application to know the chunk size in order to use this property.

Reply:
The motivation for being able to query the chunk size is due to the fact that implementations will round down the threshold or round up the maximum size, so if the user can check for this size, then they can choose more optimal sizes for these parameters.